fpsd/features.py

```diff
@@ -0,0 +1,768 @@
+def compute_bursts(df):
+    """Compute list of burst lengths for a given trace for both incoming
+    and outgoing bursts
+    """
+
+    bursts = []
+    current_incoming_burst_length = 0
+    current_outgoing_burst_length = 0
+
+    for row in df.itertuples():
+        if row.ingoing == False:
+            current_outgoing_burst_length += 1
+        elif row.ingoing == True and current_outgoing_burst_length > 0:
+            # then an outgoing burst just ended!
+            bursts.append(current_outgoing_burst_length)
+            # Reset this burst length
+            current_outgoing_burst_length = 0
+
+        if row.ingoing == True:
+            current_incoming_burst_length += 1
+        elif row.ingoing == False and current_incoming_burst_length > 0:
+            # then an incoming burst just ended!
+            bursts.append(current_incoming_burst_length)
+            # Reset this burst length
+            current_incoming_burst_length = 0
+
+    # If we were in a burst at the end of the trace, let's save it
+    if current_incoming_burst_length > 0:
+        bursts.append(current_incoming_burst_length)
+    elif current_outgoing_burst_length > 0:
+        bursts.append(current_outgoing_burst_length)
+
+    # For some features we need to know the positions of bursts,
+    # so let's save this information as well
+    ranks = list(range(1, len(bursts) + 1))
+    return bursts, ranks
+
+
+class FeatureStorage():
+    def __init__(self):
+        """Set up database engine"""
+        try:
+            self.engine = create_engine(
+                'postgresql://{}:@{}/{}'.format(
+                    *[os.environ[i] for i in
+                      ["PGUSER", "PGHOST", "PGDATABASE"]]))
+        except KeyError as exc:
+            panic("The following env vars must be set in order to know which "
+                  "database to connect to: PGUSER, PGHOST, & PGDATABASE."
+                  "\n{}.".format(exc))
+        except OperationalError as exc:
+            panic("FingerprintSecureDrop Postgres support relies on use of a "
+                  "PGPASSFILE. Make sure this file and the env var pointing "
+                  "to it exist and set 0600 permissions & user ownership."
+                  "\n{}.".format(exc))
+
+    def drop_table(self, table_name):
+        """Try to remove a table even if views depend on it
+
+        Args:
+            table_name [string]: Name of table to be dropped
+        """
+
+        query = "DROP TABLE IF EXISTS {} CASCADE;".format(table_name)
+        self.engine.execute(query)
+
+    def generate_table_undefended_frontpage_links(self):
+        """This method creates a table of exampleids that were present 
+        at the time of feature generation that the unified view can join
+        on. The table contains one integer column, exampleid.
+        """
+
+        self.drop_table("features.undefended_frontpage_examples")
+
+        query = ("CREATE TABLE features.undefended_frontpage_examples AS ( "
```

Comment:
ya this is intentional: for whatever reason rarely the crawler occasionally but rarely puts traces with no cells in the database (probably when it crashes). Those need to get filtered out but I really don’t think it’s worth messing about with the crawler for this so I just made this column to join against that selects those traces that have one or more cells